ntp-daemon/src/system.rs

```diff
@@ -210,32 +184,93 @@ impl<C: NtpClock> System<C> {
+    }
+
+    async fn handle_peer_update(&mut self, msg: MsgForSystem) {
+        tracing::debug!(?msg, "updating peer");
+
+        match msg {
+            MsgForSystem::MustDemobilize(index) => {
+                self.handle_peer_demobilize(index);
+            }
+            MsgForSystem::NewMeasurement(index, snapshot, measurement, packet) => {
+                self.handle_peer_measurement(index, snapshot, measurement, packet);
+            }
+            MsgForSystem::UpdatedSnapshot(index, snapshot) => {
+                self.handle_peer_snapshot(index, snapshot);
+            }
+            MsgForSystem::NetworkIssue(index) => {
+                self.handle_peer_network_issue(index).await;
+            }
+        }
+
         // Don't care if there is no receiver for peer snapshots (which might happen if
         // we don't enable observing in the configuration)
         let _ = self
             .peer_snapshots_sender
             .send(self.observe_peers().collect());
-        let _ = self.server_data_sender.send(self.servers().collect());
     }
 
-    #[cfg(test)]
-    fn create_test_peer(&mut self, addr: NormalizedAddress) -> PeerIndex {
-        let index = self.peer_indexer.get();
+    async fn handle_peer_network_issue(&mut self, index: PeerIndex) {
+        // Restart the peer reusing its configuration.
+        let config = self.peers.remove(&index).unwrap().peer_address;
+        match config {
+            PeerAddress::Peer { address } => {
+                self.add_peer_internal(address).await;
+            }
+            PeerAddress::Pool {
+                index,
+                address,
+                max_peers,
+                ..
+            } => {
+                self.add_to_pool(index, address, max_peers).await;
+            }
+        }
+    }
 
-        self.peers.insert(
+    fn handle_peer_snapshot(&mut self, index: PeerIndex, snapshot: PeerSnapshot) {
+        self.controller.peer_update(
             index,
-            PeerState {
-                snapshot: None,
-                peer_address: PeerAddress::Peer { address: addr },
-            },
+            snapshot
+                .accept_synchronization(self.config.local_stratum)
+                .is_ok(),
         );
-        self.controller.peer_add(index);
+        self.peers.get_mut(&index).unwrap().snapshot = Some(snapshot);
+    }
 
-        index
+    fn handle_peer_measurement(
+        &mut self,
+        index: PeerIndex,
+        snapshot: PeerSnapshot,
+        measurement: ntp_proto::Measurement,
+        packet: ntp_proto::NtpPacket<'static>,
+    ) {
+        self.handle_peer_snapshot(index, snapshot);
+        let result = self.controller.peer_measurement(index, measurement, packet);
+        if let Some((used_peers, timedata)) = result {
+            self.system.update(
+                used_peers
+                    .iter()
+                    .map(|v| self.peers.get(v).and_then(|data| data.snapshot).unwrap()),
```

Comment:
is `unwrap`  the right call here (`filter_map` would be another option)

Reply:
It should be a critical error if a peer is used for synchronization that system doesn't know anything about, hence the unwrap.

You are right it is a bit awkward though, I'll make it an expect that also conveys a bit more about what went wrong.